### PR TITLE
feat!: add computed members support

### PR DESCRIPTION
**Pull request by switchdk**
- [x] Add complete test suite
- [x] Baseline module with template
- [x] Implement `computed_members`
- [x] Update documentation to reflect the latest updates
- [x] Update Changelog

### CODE REVIEW

File: README.tfdoc.hcl [hunk 1]
```diff
@@ -142,13 +118,22 @@ section {
...
+           A map of members to replace in `members` of various IAM settings to handle terraform computed values.
+         END
+          default     = {}
+        }
+
         variable "role" {
           type        = string
           description = <<-END
-            The role that should be applied. Note that custom roles must be of the format `[projects|organizations]/{parent-name}/roles/{role-name}`.
+            The role that should be applied. Note that custom roles must be of the format `[projects|organizations]/{parent-name}/roles/{role-name}`. Cannot be used at the same time as `policy_bindings`.
```
Comment (waxb) [inline README.tfdoc.hcl:136]:
```suggestion
            The role that should be applied. Note that custom roles must be of the format `[projects|organizations]/{parent-name}/roles/{role-name}`. Omitted if `policy_bindings` set.
```

File: outputs.tf [hunk 2]
```diff
@@ -1,15 +1,12 @@
...
+  binding          = one(google_cloud_run_service_iam_binding.binding)
+  member           = google_cloud_run_service_iam_member.member
+  policy           = one(google_cloud_run_service_iam_policy.policy)
+  iam_output       = [local.binding, local.member, local.policy]
   iam_output_index = var.policy_bindings != null ? 2 : var.authoritative ? 0 : 1
 }
 
 output "iam" {
-  description = "All attributes of the created 'google_cloud_run_service_iam_binding' or 'google_cloud_run_service_iam_member' or 'google_cloud_run_service_iam_policy' resource according to the mode."
+  description = "All attributes of the created `iam_binding` or `iam_member` or `iam_policy` resource according to the mode."
```
Comment (waxb) [inline outputs.tf:10]: why did you strip the resource name?

File: test/unit-complete/main.tf [hunk 3]
```diff
@@ -1,13 +1,52 @@
...
+    "projectEditor:${local.project_id}",
+    "projectViewer:${local.project_id}",
+    "computed:service-account-name@${local.project_id}.iam.gserviceaccount.com"
+  ]
+
+  computed_members_map = {
+    myserviceaccount = "serviceAccount:${module.test-sa.service_account.email}"
+  }
 
-  project = local.project_id
```
Comment (waxb) [inline test/unit-complete/main.tf:9]: why is this removed?

File: test/unit-complete/main.tf [hunk 4]
```diff
@@ -18,7 +57,48 @@ module "test" {
...
+  location = var.gcp_region
+
+  # add all optional arguments that create additional/extended resources
+  computed_members_map = {
+    myserviceaccount = "serviceAccount:${module.test-sa.service_account.email}"
+  }
+
+  role = null # default
+
+  authoritative = true # default
```
Comment (waxb) [inline test/unit-complete/main.tf:91]:
```suggestion
```

File: test/unit-disabled/main.tf [hunk 5]
```diff
@@ -1,12 +1,58 @@
-module "test" {
+module "test1" {
   source = "../.."
 
+  # add all required arguments
```
Comment (waxb) [inline test/unit-disabled/main.tf:4]: i think the comment here was misleading, we usually do not add all arguments to `unit-disabled`, just the required ones, testing if `disabled` works.

File: variables.tf [hunk 6]
```diff
@@ -60,12 +71,12 @@ variable "policy_bindings" {
 
 variable "module_enabled" {
   type        = bool
-  description = "(Optional) Whether to create resources within the module or not. Default is 'true'."
+  description = "(Optional) Whether or not to create resources within the module."
```
Comment (waxb) [inline variables.tf:74]:
```suggestion
  description = "(Optional) Toggle resource creation within the module"
```

**Review by waxb — CHANGES_REQUESTED** [review]

**Comment by waxb** [comment]
please add [terramate manifest](https://github.com/mineiros-io/terraform-module-template/tree/main/.terramate) as well

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -20,7 +20,7 @@ For bug reports or requests, please submit your issue in the appropriate reposit
 We advise that you open an issue and ask the
 [CODEOWNERS] and community prior to starting a contribution.
 This is your chance to ask questions and receive feedback before
```
[the remark this answers is not in the document]
Comment (switchdk) [reply CONTRIBUTING.md:22]: I believe this is already in the template https://github.com/mineiros-io/terraform-module-template/blob/main/CONTRIBUTING.md?plain=1#L20-L24. Maybe I have missed something.

File: CHANGELOG.md [hunk 8]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Support for computed members
+
+### Changed
```
[the remark this answers is not in the document]
Comment (switchdk) [reply CHANGELOG.md:13]: fixed

File: Makefile [hunk 9]
```diff
@@ -124,6 +124,11 @@ test/unit-tests:
 terradoc:
 	$(call quiet-command,terradoc generate -o README.md README.tfdoc.hcl)
 
+## Generate shared configuration for tests
```
[the remark this answers is not in the document]
Comment (switchdk) [reply Makefile:127]: I don't believe I have made any changes to the `Makefile` at all. I copied it from the `main` branch from the `terraform-module-template` repo. Here is the line in question but maybe there is something else I have missed https://github.com/mineiros-io/terraform-module-template/blob/main/Makefile#L127

File: README.tfdoc.hcl (hunk 1 above)
Comment (switchdk) [reply README.tfdoc.hcl:136]: Has been changed to 
```suggestion
            The role that should be applied. Note that custom roles must be of the format `[projects|organizations]/{parent-name}/roles/{role-name}`. Omit if `policy_bindings` is set.
```

File: test/unit-complete/main.tf (hunk 4 above)
Comment (switchdk) [reply test/unit-complete/main.tf:91]: I am curious why we do not want the default values stated explicitly in the `unit-complete`. I think it is good to show all values that are being tested in `unit-complete`. For other tests, I don't think it is relevant.

File: test/unit-disabled/main.tf (hunk 5 above)
Comment (switchdk) [reply test/unit-disabled/main.tf:4]: that was just a `paste` mistake. Forgot to remove it

Comment (switchdk) [reply test/unit-disabled/main.tf:4]: Does your comment relate to the entire file i.e. there should only be a single test with the two required arguments? I was inspired by this [test/unit-disabled/main.tf](https://github.com/mineiros-io/terraform-google-service-account-iam/blame/main/test/unit-disabled/main.tf)

File: variables.tf (hunk 6 above)
Comment (switchdk) [reply variables.tf:74]: Has been changed to
```suggestion
  description = "(Optional) Toggle resource creation within the module."
```

File: outputs.tf (hunk 2 above)
Comment (switchdk) [reply outputs.tf:10]: Because that is how it is done in [terraform-google-service-account-iam](https://github.com/mineiros-io/terraform-google-service-account-iam/blob/main/outputs.tf#L12) and [terraform-google-storage-bucket-iam](https://github.com/mineiros-io/terraform-google-storage-bucket-iam/blob/main/outputs.tf#L12)

Comment (switchdk) [reply outputs.tf:10]: Ok, thanks for the explanation (separately). I will revert this change and then over time, we will have to address this in the other repos.

File: variables.tf (hunk 6 above)
Comment (switchdk) [reply variables.tf:74]: @switchdk remember to backport to the template.

File: test/unit-disabled/main.tf (hunk 5 above)
Comment (switchdk) [reply test/unit-disabled/main.tf:4]: Thanks for the explanation (and discussion separately). Let us adhere to the template test module structure and comments in the future.

File: test/unit-complete/main.tf (hunk 4 above)
Comment (switchdk) [reply test/unit-complete/main.tf:91]: Thanks for the explanation. The idea is by not stating the defaults in the tests is that you can just refer to the documentation. It is not necessary to be able to read the test without referring to the documentation.

File: test/unit-complete/main.tf (hunk 3 above)
Comment (switchdk) [reply test/unit-complete/main.tf:9]: A mistake, I will revert this change.

Comment (switchdk) [reply test/unit-complete/main.tf:9]: fixed.

File: CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Test suite to be complete
+
```
Comment (mariux) [inline CHANGELOG.md:15]: we are missing the breking changes of dropping support for previous major terraform and provider versions..

please also mark commits and PR as breaking using conventional commits `!` marker

**Review by mariux — CHANGES_REQUESTED** [review]

File: outputs.tf (hunk 2 above)
Comment (mariux) [reply outputs.tf:10]: both options would be fine for me tbh.. your choice ;)

File: CHANGELOG.md (hunk 10 above)
Comment (switchdk) [reply CHANGELOG.md:15]: Thanks @mariux,  Suggestions have been addressed.

**Review by soerenmartius — DISMISSED** [review]
lgtm!

File: CHANGELOG.md [hunk 11]
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Support for computed members
```
Comment (waxb) [inline CHANGELOG.md:11]:
```suggestion
- Add support for computed members
```

**Review by waxb — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 11 above)
Comment (switchdk) [reply CHANGELOG.md:11]: I appreciate you would like to include the word `Add` but I intentionally didn't include it since the heading of the section is `Added`. So, if you read the heading + bullet point, it reads: Added support for computed members :grin:

Comment (waxb) [reply CHANGELOG.md:11]: this is @mariux's convention, check any changelog in the past half year.
and also, i would agree and accept your stance if you would do this with all the other [sections](https://github.com/mineiros-io/terraform-google-cloud-run-iam/blob/71ed7c2bea762dcb248b89095197f388345928b8/CHANGELOG.md?plain=1#L14) as well :)

File: CHANGELOG.md [hunk 12]
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Removed
+- BREAKING: Remove support for Terraform versions before v1.0
+- BREAKING: Remove support for Terraform Google Provider version 3.x
```
Comment (waxb) [inline CHANGELOG.md:15]:
```suggestion
- BREAKING: Support for Terraform versions before v1.0
- BREAKING: Support for Terraform Google Provider version 3.x
```

makes no sense :)

**Review by waxb — CHANGES_REQUESTED** [review]

File: CHANGELOG.md (hunk 11 above)
Comment (switchdk) [reply CHANGELOG.md:11]: hahaha, nice catch. I failed at being consistent.

**Review by waxb — APPROVED** [review]
lgtm, great job, lets do the release party

**Review by mariux — APPROVED** [review]
lgtm
